compiler/crates/relay-lsp/src/inlay_hints.rs

```diff
@@ -0,0 +1,184 @@
+/*
+ * Copyright (c) Meta Platforms, Inc. and affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+use std::path::PathBuf;
+
+use common::Location;
+use common::SourceLocationKey;
+use common::Span;
+use graphql_ir::Argument;
+use graphql_ir::FragmentDefinition;
+use graphql_ir::FragmentSpread;
+use graphql_ir::InlineFragment;
+use graphql_ir::OperationDefinition;
+use graphql_ir::Program;
+use graphql_ir::Visitor;
+use intern::string_key::Intern;
+use intern::string_key::StringKey;
+use lsp_types::request::InlayHintRequest;
+use lsp_types::request::Request;
+use lsp_types::InlayHint;
+use lsp_types::InlayHintLabel;
+use lsp_types::InlayHintTooltip;
+use lsp_types::MarkupContent;
+use schema::Schema;
+
+use crate::location::transform_relay_location_to_lsp_location;
+use crate::lsp_runtime_error::LSPRuntimeResult;
+use crate::server::GlobalState;
+use crate::utils::is_file_uri_in_dir;
+use crate::LSPRuntimeError;
+
+pub fn on_inlay_hint_request(
+    state: &impl GlobalState,
+    params: <InlayHintRequest as Request>::Params,
+) -> LSPRuntimeResult<<InlayHintRequest as Request>::Result> {
+    let uri = params.text_document.uri.clone();
+    let root_dir = state.root_dir();
+
+    if !is_file_uri_in_dir(&root_dir, &uri) {
+        return Err(LSPRuntimeError::ExpectedError);
+    }
+
+    let absolute_file_path = uri.to_file_path().map_err(|_| {
+        LSPRuntimeError::UnexpectedError(format!("Unable to convert URL to file path: {:?}", uri))
+    })?;
+
+    let file_path = absolute_file_path.strip_prefix(&root_dir).map_err(|_e| {
+        LSPRuntimeError::UnexpectedError(format!(
+            "Failed to strip prefix {:?} from {:?}",
+            root_dir, absolute_file_path
+        ))
+    })?;
+
+    let project_name = state.extract_project_name_from_url(&uri)?;
+    let program = state.get_program(&project_name)?;
```

Comment:
Using this program's IR as the source of truth here will not work. This is based on what's on disk and not what's currently open in the editor. It will not reflect unsaved changes.

https://github.com/facebook/relay/assets/162735/66b21355-1c91-44be-ba26-4f52ba645694

Reply:
This should do the trick: https://github.com/facebook/relay/pull/4509

Reply:
Yes! I ended up discovering that the hard way. I needed that change as well as figuring out how to derive the IR from the cached changed file. I did a hacky version of your change here, but let me see if I can do it via rebase and (finally) get your pr merged.